### DISCUSSION TITLE
On the future functions of `Enumerable`

### DISCUSSION THREAD

**Discussion by aarkegz** [GeminiLab/enumerable#38] [Ideas]
Mathematically speaking, function `enumerator` is enough for everything, since developers can just `collect` the results to any container they prefer and do any operation they want. But in practice, it is definitely not enough. Sampling, for example, is a very common operation that is not possible to be done efficiently with just `enumerator`. Some may also want map values from/to their indexes of enumerating, which is also slow with just `enumerator`.

It's just like `std::iter::Iterator`, `next` is theoretically the only method needed, but 75 other methods are there for better performance and convenience (by the way, the derived enumerators for `enum`s and `struct`s should implement some of them too, I'll talk about it later in another session).

I'd like to list some of the functions that I believe are worth adding to `Enumerable`, and obstacles that may be encountered when implementing them.

## Sampling

Sampling is quite simple to implement on almost everything:

- Built-in numeric types, `bool`, and `char`: `rand` has done this for us,
- Tuples, `struct`s, and fields of variants of `enum`s: just sample each field,
- Plain `enum`s: sample each variant with equal probability,
- `()`: always return `()`.

The only problem is enums with fields, to sample all possible values with equal probability, we need to know the number of possible values of each variant, and pray that the sum is not too large, or uniform sampling is not possible.

A possible workaround is to sample all variants with equal probability, and then sample fields. This works for all enums, with the only exception that variants with uninhabited fields should be ignored (it's annoying but still able to implement). Helper attributes may also be introduced to allow developers to specify the distribution they want.

```rust
#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = by_variant)]
enum Foo {
    A,
    B(u8),
    C(u8, u8),
}

#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = by_value)]
enum Bar {
    A,
    B(u8),
    C(u8, u8),
}

#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = by_chance)]
enum Baz {
    #[enumerable(chance = 1)]
    A,
    #[enumerable(chance = 10)]
    B(u8),
    #[enumerable(chance = 100)]
    C(u8, u8),
}

/// the possibilities of `A`, `B`, and `C` are all equal
let foo = Foo::sample();
/// the possibilities of all possible values are equal, so compared with `foo`,
/// `bar` is much less likely to be `A` and much more likely to be `C`
let bar = Bar::sample();
/// the possibility of `C` is 100 times of `A`, and 10 times of `B`
let baz = Baz::sample();

/// which one should be the default one?
/// > `by_value` should, as it is consistent with the default behavior of `Enumerable`
```

There is still another thing to consider: "sampling" is somewhat independent of "enumerating", which means maybe a separate trait is better for this. But an extended `Enumerable` trait should not be a problem, an inefficient default implementation is also possible.

## Mapping values from/to indexes

It won't be hard to implement, but there are some things to consider:

1. How to deal with large types (types with more than `usize::MAX` possible values)? Some large types may have their first `usize::MAX` values mapped from/to indexes easily (`u128` for example), but some may be very hard (yet possible), like:

    ```rust
    struct Large1(bool, (u128, u128));

    enum Large2 {
        Some(bool),
        Other(Large1),
        Thing(u128),
    }
    ```

    Implementing `enumerable_from_index`/`enumerable_index_of` for these types is just somehow too complex. Maybe we should let it throw a compile-time error when the type is too large (by panicking when evaluating some `const`s), as it won't be a major version change if I decide to implement it in the future (or never).

2. How to deal with types that are not `PartialEq`? It's not a problem for `enumerable_from_index`, but it is for the (critical) default implementation of `enumerable_index_of`. Types bounds on `enumerable_index_of` may be a solution, like

    ```rust
    trait Enumerable: Copy {
        // ...
        fn enumerable_index_of(value: Self) -> Option<usize> where Self: PartialEq {
            Self::enumerator().position(|x| x == *self)
        }
        // ...
    }
    ```

    But it will be a disaster for derive proc-macro, as it is completely impossible to determine whether the implementing type is `PartialEq` or not, while the auto-generated code is essential because the default implementation is just too, too slow.

    I have currently no idea how to solve this problem. Enforcing `Enumerable: PartialEq` is not a good idea, as it is not always necessary and it requires a major version change.

**Comment by aarkegz** [comment]
> ```rust
> #[derive(Copy, Clone, Enumerable)]
> #[enumerable(sample = by_variant)]
> enum Foo {
>     A,
>     B(u8),
>     C(u8, u8),
> }
> 
> #[derive(Copy, Clone, Enumerable)]
> #[enumerable(sample = by_value)]
> enum Bar {
>     A,
>     B(u8),
>     C(u8, u8),
> }
> 
> #[derive(Copy, Clone, Enumerable)]
> #[enumerable(sample = by_chance)]
> enum Baz {
>     #[enumerable(chance = 1)]
>     A,
>     #[enumerable(chance = 10)]
>     B(u8),
>     #[enumerable(chance = 100)]
>     C(u8, u8),
> }
> ```

When implementing sampling `enum`s, there's something to notice,

- Maybe `sample = by_variant/by_value/by_chance` is not clear enough, I'm considering to change it to `sample = variants/values/chances`.
- Another trait `SamplableEnum` is essential, as I want to make it possible to override the default behavior of a `enum` field, like:

    ```rust
    #[derive(Copy, Clone, Enumerable)]
    #[enumerable(sample = variants)]
    enum Foo {
        A,
        B(u8),
        C(u8, u8),
    }

    #[derive(Copy, Clone, Enumerable)]
    struct Bar {
        #[enumerable(sample = values)]
        a: Foo,
        #[enumerable(sample = chances, chances = [1, 10, 100])]
        b: Foo,
    }
    ```

**Reply by aarkegz** [reply]
> > ```rust
> > #[derive(Copy, Clone, Enumerable)]
> > #[enumerable(sample = by_variant)]
> > enum Foo {
> >     A,
> >     B(u8),
> >     C(u8, u8),
> > }
> > 
> > #[derive(Copy, Clone, Enumerable)]
> > #[enumerable(sample = by_value)]
> > enum Bar {
> >     A,
> >     B(u8),
> >     C(u8, u8),
> > }
> > 
> > #[derive(Copy, Clone, Enumerable)]
> > #[enumerable(sample = by_chance)]
> > enum Baz {
> >     #[enumerable(chance = 1)]
> >     A,
> >     #[enumerable(chance = 10)]
> >     B(u8),
> >     #[enumerable(chance = 100)]
> >     C(u8, u8),
> > }
> > ```
> 
> When implementing sampling `enum`s, there's something to notice,
> 
> * Maybe `sample = by_variant/by_value/by_chance` is not clear enough, I'm considering to change it to `sample = variants/values/chances`.
> * Another trait `SamplableEnum` is essential, as I want to make it possible to override the default behavior of a `enum` field, like:
>   ```rust
>   #[derive(Copy, Clone, Enumerable)]
>   #[enumerable(sample = variants)]
>   enum Foo {
>       A,
>       B(u8),
>       C(u8, u8),
>   }
>   
>   #[derive(Copy, Clone, Enumerable)]
>   struct Bar {
>       #[enumerable(sample = values)]
>       a: Foo,
>       #[enumerable(sample = chances, chances = [1, 10, 100])]
>       b: Foo,
>   }
>   ```

It's decided to be `sample = from_values/from_variants/with_chances`, like:

```rust
#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = from_values)]
enum Foo {
    A,
    B(u8),
    C(u8, u8),
}

#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = from_variants)]
enum Bar {
    A,
    B(u8),
    C(u8, u8),
}

#[derive(Copy, Clone, Enumerable)]
#[enumerable(sample = with_chances)]
enum Baz {
    #[enumerable(chance = 1)]
    A,
    #[enumerable(chance = 10)]
    B(u8),
    #[enumerable(chance = 100)]
    C(u8, u8),
}
```

**Comment by tyilo** [comment]
I just discovered this crate.

For my [`exhaustive-map`](https://docs.rs/exhaustive-map/) crate I have a similar trait as `Enumerable` called [`Finite`](https://docs.rs/exhaustive-map/latest/exhaustive_map/trait.Finite.html).

It has the following definition:
```rust
pub trait Finite: Sized {
    const INHABITANTS: usize;

    fn to_usize(&self) -> usize;
    fn from_usize(i: usize) -> Option<Self>;
}
```

and implements `iter_all()` (similar to `Enumerable::enumerator`) using an extension trait:
```rust
pub trait FiniteExt: Finite {
    fn iter_all() -> IterAll<Self> {
        ...
    }
}
```

Feel free to use that as inspiration.

**Reply by aarkegz** [reply]
Hello, and Thanks for your comment! Your crates is quite good and I think `Finite` and `FiniteExt` is exactly what should be added here. The question is I'm still trying to find an intuitive, efficient and backward-compatible way to handle types too large to be represented by `usize`.
